eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/DualFundingHandlers.scala

```diff
@@ -32,22 +37,56 @@ trait DualFundingHandlers extends CommonHandlers {
 
   this: Channel =>
 
-  def publishFundingTx(d: DATA_WAIT_FOR_DUAL_FUNDING_PLACEHOLDER): Unit = {
-    d.fundingTx match {
+  def publishFundingTx(fundingParams: InteractiveTxParams, fundingTx: SignedSharedTransaction): Unit = {
+    fundingTx match {
       case _: PartiallySignedSharedTransaction =>
         log.info("we haven't received remote funding signatures yet: we cannot publish the funding transaction but our peer should publish it")
       case fundingTx: FullySignedSharedTransaction =>
         // Note that we don't use wallet.commit because we don't want to rollback on failure, since our peer may be able
         // to publish and we may be able to RBF.
         wallet.publishTransaction(fundingTx.signedTx).onComplete {
           case Success(_) =>
-            context.system.eventStream.publish(TransactionPublished(d.commitments.channelId, remoteNodeId, fundingTx.signedTx, fundingTx.tx.localFees(d.fundingParams), "funding"))
-            channelOpenReplyToUser(Right(ChannelOpenResponse.ChannelOpened(d.commitments.channelId)))
+            context.system.eventStream.publish(TransactionPublished(fundingParams.channelId, remoteNodeId, fundingTx.signedTx, fundingTx.tx.localFees(fundingParams), "funding"))
+            channelOpenReplyToUser(Right(ChannelOpenResponse.ChannelOpened(fundingParams.channelId)))
           case Failure(t) =>
             channelOpenReplyToUser(Left(LocalError(t)))
             log.warning("error while publishing funding tx: {}", t.getMessage) // tx may be published by our peer, we can't fail-fast
         }
     }
   }
 
+  def handleNewBlockDualFundingUnconfirmed(c: CurrentBlockHeight, d: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED) = {
+    if (Channel.FUNDING_TIMEOUT_FUNDEE < c.blockHeight - d.waitingSince && Closing.nothingAtStake(d)) {
+      log.warning("funding transaction did not confirm in {} blocks and we have nothing at stake, forgetting channel", Channel.FUNDING_TIMEOUT_FUNDEE)
+      handleFundingTimeout(d)
+    } else if (d.lastChecked + 6 < c.blockHeight) {
+      log.debug("checking if funding transactions have been double-spent")
+      val fundingTxs = (d.fundingTx +: d.previousFundingTxs.map(_.fundingTx)).map(_.tx.buildUnsignedTx())
+      // We check whether *all* funding attempts have been double-spent.
+      // Since we only consider a transaction double-spent when the spending transaction is confirmed, this will not
+      // return false positives when one of our transactions is confirmed, because its individual result will be false.
+      Future.sequence(fundingTxs.map(tx => wallet.doubleSpent(tx))).map(_.forall(_ == true)).map {
+        case true => self ! BITCOIN_FUNDING_DOUBLE_SPENT(fundingTxs.map(_.txid).toSet)
+        case false => publishFundingTx(d.fundingParams, d.fundingTx) // we republish the highest feerate funding transaction
+      }
```

Comment:
> Maybe we could just decorrelate from the block, or trigger a random check every 6 blocks after a random delay (similarly to what happens in the watcher).

I'm hesitant to do that for the following reasons:

- it creates a new event that needs to be listened to, which slightly increases code complexity (not a lot though, that's not a good enough reason on its own)
- we already have a protection against herd effect since each channel randomly delays block processing to ensure we don't have all channels handling blocks at exactly the same time (see `max-block-processing-delay`)
- this check only happens while funding is unconfirmed, which in practice shouldn't last very long
- even if we have multiple channels in an unconfirmed state, they won't check for double-spend at the same block, unless they all reached the wait_for_unconfirmed state at the same moment (or at the same multiple of 6)
- I arbitrarily chose 6 blocks, but we can check that less often if you'd like, a few times a day is likely enough

Let me know if you think it's worth doing regardless of these points and I'll do it.

Reply:
> - we already have a protection against herd effect since each channel randomly delays block processing to ensure we don't have all channels handling blocks at exactly the same time (see max-block-processing-delay)

Fair points! Especially this one. Happy to leave it as-is.